simperby/src/node.rs

```diff
@@ -0,0 +1,197 @@
+mod storage;
+
+use super::*;
+use simperby_kv_storage::KVStorage;
+use simperby_network::AuthorizedNetwork;
+use std::sync::Arc;
+use storage::*;
+use tokio::sync::RwLock;
+use vetomint::Round;
+
+pub struct Node {
+    state: Arc<RwLock<NodeState>>,
+    network_task: tokio::task::JoinHandle<()>,
+    genesis_info: GenesisInfo,
+}
+
+impl Node {
+    pub(crate) async fn new(
+        genesis_info: GenesisInfo,
+        network: Box<dyn AuthorizedNetwork>,
+        state_storage: Box<dyn KVStorage>,
+        history_storage: Box<dyn KVStorage>,
+    ) -> Result<Self, anyhow::Error> {
+        let last_header = genesis_info.header.clone();
+        let state_ = Arc::new(RwLock::new(NodeState {
+            state_storage,
+            history_storage: HistoryStorage::new(history_storage, genesis_info.clone()).await?,
+            last_header,
+        }));
+        Ok(Node {
+            state: Arc::clone(&state_),
+            network_task: tokio::task::spawn(async move {
+                run_network_task(network, state_).await;
+            }),
+            genesis_info,
+        })
+    }
+}
+
+impl Drop for Node {
+    fn drop(&mut self) {
+        self.network_task.abort();
+    }
+}
+
+#[async_trait]
+impl SimperbyApi for Node {
+    fn get_genesis_info(&self) -> &GenesisInfo {
+        &self.genesis_info
+    }
+
+    async fn get_height(&self) -> u64 {
+        self.state.read().await.last_header.height
+    }
+
+    async fn get_block(&self, height: BlockHeight) -> Result<Block, SimperbyError> {
+        let state = self.state.read().await;
+        if state.last_header.height < height {
+            return Err(SimperbyError::InvalidOperation(format!(
+                "height {} is greater than the current height {}",
+                height, state.last_header.height
+            )));
+        }
+
+        state
+            .history_storage
+            .get_block(height)
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn check_block(&self, _block: Block, _height: BlockHeight) -> Result<(), SimperbyError> {
+        unimplemented!()
+    }
+
+    async fn read_state(&self, key: String, height: u64) -> Result<Vec<u8>, SimperbyError> {
+        let state = self.state.read().await;
+        assert_eq!(
+            height, state.last_header.height,
+            "currently state query is supported only for the last block"
+        );
+        state
+            .state_storage
+            .get(Hash256::hash(key))
+            .await
+            .map_err(SimperbyError::StorageError)
+    }
+
+    async fn get_consensus_vote_options(&self) -> Result<Vec<ConsensusVoteItem>, SimperbyError> {
```

Comment:
```suggestion
    async fn get_consensus_voting_options(&self) -> Result<Vec<ConsensusVoteItem>, SimperbyError> {
```
https://github.com/postech-dao/simperby/pull/51/files#r953479470
Same here.

Reply:
ditto.